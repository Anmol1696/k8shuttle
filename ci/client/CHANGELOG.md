# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.2](https://github.com/cosmology-tech/starshipjs/compare/@starship-ci/client@0.1.1...@starship-ci/client@0.1.2) (2024-04-20)

**Note:** Version bump only for package @starship-ci/client





## [0.1.1](https://github.com/cosmology-tech/starshipjs/compare/@starship-ci/client@0.1.0...@starship-ci/client@0.1.1) (2024-04-20)

**Note:** Version bump only for package @starship-ci/client





# [0.1.0](https://github.com/cosmology-tech/starshipjs/compare/@starship-ci/client@0.0.5...@starship-ci/client@0.1.0) (2024-04-20)

**Note:** Version bump only for package @starship-ci/client





## [0.0.5](https://github.com/cosmology-tech/starshipjs/compare/@starship-ci/client@0.0.4...@starship-ci/client@0.0.5) (2024-04-20)

**Note:** Version bump only for package @starship-ci/client





## [0.0.4](https://github.com/cosmology-tech/starshipjs/compare/@starship-ci/client@0.0.3...@starship-ci/client@0.0.4) (2024-04-20)

**Note:** Version bump only for package @starship-ci/client





## [0.0.3](https://github.com/cosmology-tech/starshipjs/compare/@starship-ci/client@0.0.2...@starship-ci/client@0.0.3) (2024-04-20)

**Note:** Version bump only for package @starship-ci/client





## 0.0.2 (2024-04-20)

**Note:** Version bump only for package @starship-ci/client
