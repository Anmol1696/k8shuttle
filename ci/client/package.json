{
  "name": "@starship-ci/client",
  "version": "0.1.2",
  "author": "Dan Lynch <pyramation@gmail.com>",
  "description": "Starship CI Client",
  "main": "index.js",
  "module": "esm/index.js",
  "types": "index.d.ts",
  "homepage": "https://github.com/cosmology-tech/starshipjs",
  "license": "SEE LICENSE IN LICENSE",
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/cosmology-tech/starshipjs"
  },
  "bugs": {
    "url": "https://github.com/cosmology-tech/starshipjs/issues"
  },
  "scripts": {
    "copy": "copyfiles -f ../../LICENSE README.md package.json dist",
    "clean": "del dist/**",
    "prepare": "npm run build",
    "build": "npm run clean; tsc; tsc -p tsconfig.esm.json; npm run copy",
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "keywords": [],
  "dependencies": {
    "@chain-registry/types": "^0.25.5",
    "@types/js-yaml": "^4.0.9",
    "chalk": "^4.1.0",
    "deepmerge": "^4.3.1",
    "js-yaml": "^4.1.0",
    "mkdirp": "3.0.1",
    "semver": "7.6.0",
    "shelljs": "^0.8.5"
  },
  "devDependencies": {
    "@types/shelljs": "^0.8.15"
  }
}
