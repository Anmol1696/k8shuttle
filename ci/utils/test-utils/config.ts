import { readFileSync } from 'fs';
import * as yaml from 'js-yaml';
import { join, resolve } from 'path';

const fixtureDir = resolve(join(__dirname, '/../../../__fixtures__'));

function loadConfig(filename: string) {
    const configPath = join(fixtureDir, filename);
    const configAsYaml = readFileSync(configPath, 'utf-8');
    const config = yaml.load(configAsYaml);
    return {
        configPath,
        configAsYaml,
        config
    };
}


const localConfig = loadConfig('local-config.yaml');
const config = loadConfig('config.yaml');

export { config, localConfig };
