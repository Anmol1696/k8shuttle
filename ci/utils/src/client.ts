import chalk from 'chalk';
import * as os from 'os';
import * as yaml from 'js-yaml';
import * as shell from 'shelljs';
import { Chain, StarshipConfig } from './config';
import { readFileSync } from 'fs';
import { dependencies as defaultDependencies, Dependency } from "./deps";
import { readAndParsePackageJson } from './package';
import { Ports } from './config';
export interface StarshipContext {
  helmName: string;
  helmFile: string;
  helmRepo: string;
  helmRepoUrl: string;
  helmChart: string;
  helmVersion: string;
  kindCluster?: string;
  verbose?: boolean;
  curdir?: string;
};

export interface PodPorts {
  registry: Ports,
  explorer: Ports,
  chains: {
    defaultPorts: Ports,
    [chainName: string]: Ports
  }
}

// TODO talk to Anmol about moving these into yaml, if not already possible?
const defaultPorts: PodPorts = {
  explorer: {
    rest: 8080
  },
  registry: {
    grpc: 9090,
    rest: 8080
  },
  chains: {
    defaultPorts: {
      rpc: 26657,
      rest: 1317,
      exposer: 8081,
      faucet: 8000
    }
  }
};
export interface StarshipClientI {
  ctx: StarshipContext;
  version: string;
  dependencies: Dependency[];
  depsChecked: boolean;
  config: StarshipConfig;
  podPorts: PodPorts
};

export class StarshipClient implements StarshipClientI{
  ctx: StarshipContext;
  version: string;
  dependencies: Dependency[] = defaultDependencies;
  depsChecked: boolean = false;
  config: StarshipConfig;
  podPorts: PodPorts = defaultPorts;

  constructor(ctx: StarshipContext) {
    this.ctx = ctx;
    // TODO add semver check against net
    this.version = readAndParsePackageJson().version;
  }

  private exec(cmd: string[]): shell.ShellString {
    this.checkDependencies();
    const str = cmd.join(' ');
    this.log(str);
    return shell.exec(str);
  }

  private log(str: string): void {
    // add log level
    this.log(str);
  }

  private exit(code: number): void {
    shell.exit(code);
  }

  private checkDependencies(): void {
    if (this.depsChecked) return;

    const platform = os.platform();
    const messages: string[] = [];
    const depMessages: string[] = [];
    const missingDependencies = this.dependencies.filter(dep => !dep.installed);

    if (!missingDependencies.length) {
      this.depsChecked = true;
      return;
    }

    this.dependencies.forEach(dep => {
      if (missingDependencies.find(d => d.name === dep.name)) {
        depMessages.push(`${chalk.red('x')}${dep.name}`);
      } else {
        depMessages.push(`${chalk.green('âœ“')}${dep.name}`);
      }
    });

    messages.push('\n'); // Adding a newline for better readability

    missingDependencies.forEach(dep => {
      messages.push(chalk.bold.white(dep.name + ': ') + chalk.cyan(dep.url));

      if (dep.name === 'helm' && platform === 'darwin') {
        messages.push(chalk.gray("Alternatively, you can install using brew: ") + chalk.white.bold("`brew install helm`"));
      }

      if (dep.name === 'kubectl' && platform === 'darwin') {
        messages.push(chalk.gray("Alternatively, you can install Docker for Mac which includes Kubernetes: ") + chalk.white.bold(dep.macUrl));
      }

      if (dep.name === 'docker' && platform === 'darwin') {
        messages.push(chalk.gray("For macOS, you may also consider Docker for Mac: ") + chalk.white.bold(dep.macUrl));
      } else if (dep.name === 'docker') {
        messages.push(chalk.gray("For advanced Docker usage and installation on other platforms, see: ") + chalk.white.bold("https://docs.docker.com/engine/install/"));
      }

      messages.push('\n'); // Adding a newline for separation between dependencies
    });

    this.log(depMessages.join('\n'));
    this.log('\nPlease install the missing dependencies:');
    this.log(messages.join('\n'));
    this.exit(1);
  }

  public setup(): void {
    this.setupHelm();
  }

  public teardown(): void {
    this.removeHelm();
  }

  public loadConfig(): void {
    const fileContents = readFileSync(this.ctx.helmFile, 'utf8');
    this.config = yaml.load(fileContents) as StarshipConfig;
  }

  public setConfig(config: StarshipConfig): void {
    this.config = config;
  }

  // TODO do we need this here?
  public test(): void {
    this.exec([
      'yarn',
      'run',
      'jest',
      `--testPathPattern=../${this.ctx.helmRepo}`,
      '--verbose',
      '--bail'
    ]);
  }

  public stop(): void {
    this.stopForward();
    this.deleteHelm();
  }

  public clean(): void {
    this.stop();
    this.cleanKind();
  }

  private setupHelm(): void {
    this.exec([
      'helm',
      'repo',
      'add',
      this.ctx.helmRepo,
      this.ctx.helmRepoUrl
    ]);
    this.exec(['helm', 'repo', 'update']);
    this.exec([
      'helm',
      'search',
      'repo',
      `${this.ctx.helmRepo}/${this.ctx.helmChart}`,
      '--version',
      this.ctx.helmVersion
    ]);
  }

  private removeHelm(): void {
    this.exec([
      'helm',
      'repo',
      'remove',
      this.ctx.helmRepo
    ]);
  }

  public install(): void {
    this.log("Installing the helm chart. This is going to take a while.....");
    this.exec([
      'helm',
      'install',
      '-f',
      this.ctx.helmFile,
      this.ctx.helmName,
      `${this.ctx.helmRepo}/${this.ctx.helmChart}`,
      '--version',
      this.ctx.helmVersion
    ]);
    this.log("Run \"kubectl get pods\" to check the status of the cluster");
  }

  public upgrade(): void {
    this.exec([
      'helm',
      'upgrade',
      '--debug',
      '-f',
      this.ctx.helmFile,
      this.ctx.helmName,
      `${this.ctx.helmRepo}/${this.ctx.helmChart}`,
      '--version',
      this.ctx.helmVersion
    ]);
  }

  public debug(): void {
    this.exec([
      'helm',
      'install',
      '--dry-run',
      '--debug',
      '-f',
      this.ctx.helmFile,
      this.ctx.helmName,
      `${this.ctx.helmRepo}/${this.ctx.helmChart}`
    ]);
  }

  private deleteHelm(): void {
    this.exec(['helm', 'delete', this.ctx.helmName]);
  }


  private forwardPort(chain: Chain, localPort: number, externalPort: number): void {
    if (localPort !== undefined && externalPort !== undefined) {
      this.exec([
        "kubectl", "port-forward",
        `pods/${chain.name}-genesis-0`,
        `${localPort}:${externalPort}`,
        ">", "/dev/null",
        "2>&1", "&"
      ]);
      this.log(chalk.yellow(`Forwarded ${chain.name}: local ${localPort} -> target (host) ${externalPort}`));
    }
  }
  
  private forwardPortService(serviceName: string, localPort: number, externalPort: number): void {
    if (localPort !== undefined && externalPort !== undefined) {
      this.exec([
        "kubectl", "port-forward",
        `service/${serviceName}`,
        `${localPort}:${externalPort}`,
        ">", "/dev/null",
        "2>&1", "&"
      ]);
      this.log(`Forwarded ${serviceName} on port ${localPort} to target port ${externalPort}`);
    }
  }

  public startPortForward(): void {
    if (!this.config) {
      throw new Error('no config!');
    }
    this.log("Attempting to stop any existing port-forwards...");
    this.stopPortForward();
    this.log("Starting new port forwarding...");
  
    this.config.chains.forEach(chain => {
      const chainPodPorts = this.podPorts.chains[chain.name] || this.podPorts.chains.defaultPorts;

      if (chain.ports.rpc) this.forwardPort(chain, chain.ports.rpc,  chainPodPorts.rpc);
      if (chain.ports.rest) this.forwardPort(chain, chain.ports.rest, chainPodPorts.rest);
      if (chain.ports.exposer) this.forwardPort(chain, chain.ports.exposer, chainPodPorts.exposer);
      if (chain.ports.faucet) this.forwardPort(chain, chain.ports.faucet, chainPodPorts.faucet);
    });
  
    if (this.config.registry?.enabled) {
      this.forwardPortService("registry", this.config.registry.ports.rest, this.podPorts.registry.rest);
      this.forwardPortService("registry", this.config.registry.ports.grpc, this.podPorts.registry.grpc);
    }
  
    if (this.config.explorer?.enabled) {
      this.forwardPortService("explorer", this.config.explorer.ports.rest, this.podPorts.explorer.rest);
    }
  }
  
  public stopForward(): void {
    this.exec(['pkill', '-f', 'port-forward']);
  }

  // TODO review with Anmol, which stopForward is better...
  public stopPortForward(): void {
    this.log(chalk.green("Trying to stop all port-forward, if any...."));
    const result = this.exec([
      "ps", "-ef",
      "|", "grep", "-i", "'kubectl port-forward'",
      "|", "grep", "-v", "'grep'",
      "|", "awk", "'{print $2}'"
    ]);
    const pids = (result || '').split('\n');
    pids.forEach(pid => {
      if (pid.trim()) {
        this.exec([
          "kill", "-15", pid
        ]);
      }
    });
    this.exec(['sleep', '2']);
  }

  private setupKind(): void {
    if (this.ctx.kindCluster) {
      this.exec(['kind', 'create', 'cluster', '--name', this.ctx.kindCluster]);
    }
  }

  private cleanKind(): void {
    if (this.ctx.kindCluster) {
      this.exec(['kind', 'delete', 'cluster', '--name', this.ctx.kindCluster]);
    }
  }
}
