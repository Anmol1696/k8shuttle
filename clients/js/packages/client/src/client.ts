import chalk from 'chalk';
import deepmerge from 'deepmerge';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import * as yaml from 'js-yaml';
import { mkdirp } from 'mkdirp';
import * as os from 'os';
import { dirname, resolve } from 'path';
import * as shell from 'shelljs';

import {Chain, Relayer, Script, StarshipConfig} from './config';
import { Ports } from './config';
import { dependencies as defaultDependencies, Dependency } from "./deps";
import { readAndParsePackageJson } from './package';

export interface StarshipContext {
  helmName?: string;
  helmFile?: string;
  helmRepo?: string;
  helmRepoUrl?: string;
  helmChart?: string;
  helmVersion?: string;
  helmNamespace?: string;
  verbose?: boolean;
  curdir?: string;
};

export const defaultStarshipContext: Partial<StarshipContext> = {
  helmName: 'starship',
  helmRepo: 'starship',
  helmRepoUrl: 'https://cosmology-tech.github.io/starship/',
  helmChart: 'starship/devnet',
  helmVersion: 'v0.2.4',
  helmNamespace: '',
};

export interface PodPorts {
  registry?: Ports,
  explorer?: Ports,
  chains?: {
    defaultPorts?: Ports,
    [chainName: string]: Ports
  },
  relayers?: {
    defaultPorts?: Ports,
    [relayerName: string]: Ports
  },
}

// TODO talk to Anmol about moving these into yaml, if not already possible?
const defaultPorts: PodPorts = {
  explorer: {
    rest: 8080
  },
  registry: {
    grpc: 9090,
    rest: 8080
  },
  chains: {
    defaultPorts: {
      rpc: 26657,
      grpc: 9090,
      rest: 1317,
      exposer: 8081,
      faucet: 8000,
      cometmock: 22331,
    }
  },
  relayers: {
    defaultPorts: {
      rest: 3000,
      exposer: 8081,
    }
  }
};
export interface StarshipClientI {
  ctx: StarshipContext;
  version: string;
  dependencies: Dependency[];
  depsChecked: boolean;
  config: StarshipConfig;
  podPorts: PodPorts
};

export interface PodStatus {
  phase: string;
  ready: boolean;
  restartCount: number;
  reason?: string;
}

export const formatChainID = (input: string): string => {
  // Replace underscores with hyphens
  let formattedName = input.replace(/_/g, '-');

  // Truncate the string to a maximum length of 63 characters
  if (formattedName.length > 63) {
    formattedName = formattedName.substring(0, 63);
  }

  return formattedName;
};

export class StarshipClient implements StarshipClientI {
  ctx: StarshipContext;
  version: string;
  dependencies: Dependency[] = defaultDependencies;
  depsChecked: boolean = false;
  config: StarshipConfig;
  podPorts: PodPorts = defaultPorts;

  private podStatuses = new Map<string, PodStatus>(); // To keep track of pod statuses

  // Define a constant for the restart threshold
  private readonly RESTART_THRESHOLD = 4;

  constructor(ctx: StarshipContext) {
    this.ctx = deepmerge(defaultStarshipContext, ctx);
    // TODO add semver check against net
    this.version = readAndParsePackageJson().version;
  }

  private exec(cmd: string[], log: boolean = true, silent: boolean = false): shell.ShellString {
    this.checkDependencies();
    const str = cmd.join(' ');
    if (log) this.log(str);
    return shell.exec(str, { silent });
  }

  private log(str: string): void {
    // add log level
    console.log(str);
  }

  private exit(code: number): void {
    shell.exit(code);
  }

  private checkDependencies(): void {
    if (this.depsChecked) return;

    // so CI/CD and local dev work nicely
    const platform = process.env.NODE_ENV === 'test' ? 'linux' : os.platform();
    const messages: string[] = [];
    const depMessages: string[] = [];
    const missingDependencies = this.dependencies.filter(dep => !dep.installed);

    if (!missingDependencies.length) {
      this.depsChecked = true;
      return;
    }

    this.dependencies.forEach(dep => {
      if (missingDependencies.find(d => d.name === dep.name)) {
        depMessages.push(`${chalk.red('x')}${dep.name}`);
      } else {
        depMessages.push(`${chalk.green('âœ“')}${dep.name}`);
      }
    });

    messages.push('\n'); // Adding a newline for better readability

    missingDependencies.forEach(dep => {
      messages.push(chalk.bold.white(dep.name + ': ') + chalk.cyan(dep.url));

      if (dep.name === 'helm' && platform === 'darwin') {
        messages.push(chalk.gray("Alternatively, you can install using brew: ") + chalk.white.bold("`brew install helm`"));
      }

      if (dep.name === 'kubectl' && platform === 'darwin') {
        messages.push(chalk.gray("Alternatively, you can install Docker for Mac which includes Kubernetes: ") + chalk.white.bold(dep.macUrl));
      }

      if (dep.name === 'docker' && platform === 'darwin') {
        messages.push(chalk.gray("For macOS, you may also consider Docker for Mac: ") + chalk.white.bold(dep.macUrl));
      } else if (dep.name === 'docker') {
        messages.push(chalk.gray("For advanced Docker usage and installation on other platforms, see: ") + chalk.white.bold(dep.url));
      }

      messages.push('\n'); // Adding a newline for separation between dependencies
    });

    this.log(depMessages.join('\n'));
    this.log('\nPlease install the missing dependencies:');
    this.log(messages.join('\n'));
    this.exit(1);
  }

  public setup(): void {
    this.setupHelm();
  }

  private loadYaml(filename: string): any {
    const path = filename.startsWith('/') ? filename : resolve((process.cwd(), filename))
    const fileContents = readFileSync(path, 'utf8');
    return yaml.load(fileContents);
  }

  private saveYaml(filename: string, obj: any): any {
    const path = filename.startsWith('/') ? filename : resolve((process.cwd(), filename))
    const yamlContent = yaml.dump(obj);
    mkdirp.sync(dirname(path));
    writeFileSync(path, yamlContent, 'utf8');
  }

  public loadConfig(): void {
    this.ensureFileExists(this.ctx.helmFile);
    this.config = this.loadYaml(this.ctx.helmFile) as StarshipConfig;
  }

  public saveConfig(): void {
    this.saveYaml(this.ctx.helmFile, this.config);
  }

  public savePodPorts(filename: string): void {
    this.saveYaml(filename, this.podPorts);
  }

  public loadPodPorts(filename: string): void {
    this.ensureFileExists(filename);
    this.podPorts = this.loadYaml(filename) as PodPorts;
  }

  public setConfig(config: StarshipConfig): void {
    this.config = config;
  }

  public setContext(ctx: StarshipContext): void {
    this.ctx = ctx;
  }

  public setPodPorts(ports: PodPorts): void {
    this.podPorts = deepmerge(defaultPorts, ports);
  }

  public getArgs(): string[] {
    const args = [];
    if (this.ctx.helmNamespace) {
      args.push('--namespace', this.ctx.helmNamespace);
    }
    return args;
  }

  public getDeployArgs(): string[] {
    const args = this.getArgs();
    if (this.ctx.helmNamespace) {
      args.push('--create-namespace');
    }
    return args;
  }

  // TODO do we need this here?
  public test(): void {
    this.exec([
      'yarn',
      'run',
      'jest',
      `--testPathPattern=../${this.ctx.helmRepo}`,
      '--verbose',
      '--bail'
    ]);
  }

  public stop(): void {
    this.stopPortForward();
    this.deleteHelm();
  }

  public async start(): Promise<void> {
    this.setup();
    this.deploy();
    await this.waitForPods(); // Ensure waitForPods completes before starting port forwarding
    this.startPortForward();
  }

  public setupHelm(): void {
    this.exec([
      'helm',
      'repo',
      'add',
      this.ctx.helmRepo,
      this.ctx.helmRepoUrl
    ]);
    this.exec(['helm', 'repo', 'update']);
    this.exec([
      'helm',
      'search',
      'repo',
      this.ctx.helmChart,
      '--version',
      this.ctx.helmVersion
    ]);
  }

  private ensureFileExists(filename: string): void {
    const path = filename.startsWith('/') ? filename : resolve((process.cwd(), filename))
    if (!existsSync(path)) {
      throw new Error(`Configuration file not found: ${filename}`);
    }
  }

  public deploy(options: string[] = []): void {
    this.ensureFileExists(this.ctx.helmFile);
    this.log("Installing the helm chart. This is going to take a while.....");

    const cmd: string[] = [
      'helm',
      'install',
      '-f',
      this.ctx.helmFile,
      this.ctx.helmName,
      this.ctx.helmChart,
      '--version',
      this.ctx.helmVersion,
      ...this.getDeployArgs(),
      ...options,
    ];

    // Determine the data directory of the config file
    const datadir = resolve(dirname(this.ctx.helmFile!));

    // Iterate through each chain to add script arguments
    this.config.chains.forEach((chain, chainIndex) => {
      if (chain.scripts) {
        Object.keys(chain.scripts).forEach(scriptKey => {
          const script = chain.scripts?.[scriptKey as keyof Chain['scripts']];
          if (script && script.file) {
            const scriptPath = resolve(datadir, script.file);
            cmd.push(`--set-file chains[${chainIndex}].scripts.${scriptKey}.data=${scriptPath}`);
          }
        });
      }
    });

    this.exec(cmd);
    this.log("Run \"starship get-pods\" to check the status of the cluster");
  }

  public debug(): void {
    this.ensureFileExists(this.ctx.helmFile);
    this.deploy(['--dry-run', '--debug']);
  }

  public deleteHelm(): void {
    this.exec(['helm', 'delete', this.ctx.helmName]);
  }

  public getPods(): void {
    this.exec([
      "kubectl",
      "get pods",
      ...this.getArgs(),
      // "--all-namespaces"
    ]);
  }

  private getPodNames(): string[] {
    const result = this.exec([
      'kubectl',
      'get',
      'pods',
      '--no-headers',
      '-o',
      'custom-columns=:metadata.name',
      ...this.getArgs(),
    ], false, true)
  
    // Split the output by new lines and filter out any empty lines
    const podNames = result.split('\n').filter(name => name.trim() !== '');
  
    return podNames;
  }

  public areAllPodsRunning(): boolean {
    let allRunning = true;
    this.podStatuses.forEach((status) => {
      if (status.phase !== 'Running' || !status.ready) {
        allRunning = false;
      }
    });
    return allRunning;
  }

  private checkPodStatus(podName: string): void {
    const result = this.exec([
      'kubectl',
      'get',
      'pods',
      podName,
      '--no-headers',
      '-o',
      'custom-columns=:status.phase,:status.containerStatuses[*].ready,:status.containerStatuses[*].restartCount,:status.containerStatuses[*].state.waiting.reason',
      ...this.getArgs(),
    ], false, true).trim();

    const [phase, readyList, restartCountList, reason] = result.split(/\s+/);
    const ready = readyList.split(',').every(state => state === 'true');
    const restarts = restartCountList.split(',').reduce((acc, count) => acc + parseInt(count, 10), 0);

    this.podStatuses.set(podName, { phase, ready, restartCount: restarts, reason });

    if (restarts > this.RESTART_THRESHOLD) {
      this.log(`${chalk.red('Error:')} Pod ${podName} has restarted more than ${this.RESTART_THRESHOLD} times.`);
      this.exit(1);
    }
  }

  public async waitForPods(): Promise<void> {
    const podNames = this.getPodNames();

    // Check the status of each pod retrieved
    podNames.forEach(podName => {
      this.checkPodStatus(podName);
    });

    this.displayPodStatuses();

    if (!this.areAllPodsRunning()) {
      await new Promise(resolve => setTimeout(resolve, 2500));
      await this.waitForPods(); // Recursive call
    } else {
      this.log(chalk.green('All pods are running!'));
      // once the pods are in running state, wait for 10 more seconds
      await new Promise(resolve => setTimeout(resolve, 5000));
    }
  }

  private displayPodStatuses(): void {
    console.clear();
    this.podStatuses.forEach((status, podName) => {
      let statusColor;
      if (status.phase === 'Running' && status.ready) {
        statusColor = chalk.green(status.phase);
      } else if (status.phase === 'Running' && !status.ready) {
        statusColor = chalk.yellow('RunningButNotReady');
      } else if (status.phase === 'Terminating') {
        statusColor = chalk.gray(status.phase);
      } else {
        statusColor = chalk.red(status.phase);
      }

      console.log(`[${chalk.blue(podName)}]: ${statusColor}, ${chalk.gray(`Ready: ${status.ready}, Restarts: ${status.restartCount}`)}`);
    });
  }

  private forwardPort(chain: Chain, localPort: number, externalPort: number): void {
    if (localPort !== undefined && externalPort !== undefined) {
      this.exec([
        "kubectl", "port-forward",
        `pods/${formatChainID(chain.id)}-genesis-0`,
        `${localPort}:${externalPort}`,
        ...this.getArgs(),
        ">", "/dev/null",
        "2>&1", "&"
      ]);
      this.log(chalk.yellow(`Forwarded ${formatChainID(chain.id)}: local ${localPort} -> target (host) ${externalPort}`));
    }
  }

  private forwardPortCometmock(chain: Chain, localPort: number, externalPort: number): void {
    if (localPort !== undefined && externalPort !== undefined) {
      this.exec([
        "kubectl", "port-forward",
        `pods/${formatChainID(chain.id)}-cometmock-0`,
        `${localPort}:${externalPort}`,
        ...this.getArgs(),
        ">", "/dev/null",
        "2>&1", "&"
      ]);
      this.log(chalk.yellow(`Forwarded ${formatChainID(chain.id)}: local ${localPort} -> target (host) ${externalPort}`));
    }
  }

  private forwardPortRelayer(relayer: Relayer, localPort: number, externalPort: number): void {
    if (localPort !== undefined && externalPort !== undefined) {
      this.exec([
        "kubectl", "port-forward",
        `pods/${relayer.type}-${relayer.name}-0`,
        `${localPort}:${externalPort}`,
        ...this.getArgs(),
        ">", "/dev/null",
        "2>&1", "&"
      ]);
      this.log(chalk.yellow(`Forwarded ${relayer.name}: local ${localPort} -> target (host) ${externalPort}`));
    }
  }

  private forwardPortService(serviceName: string, localPort: number, externalPort: number): void {
    if (localPort !== undefined && externalPort !== undefined) {
      this.exec([
        "kubectl", "port-forward",
        `service/${serviceName}`,
        `${localPort}:${externalPort}`,
        ...this.getArgs(),
        ">", "/dev/null",
        "2>&1", "&"
      ]);
      this.log(`Forwarded ${serviceName} on port ${localPort} to target port ${externalPort}`);
    }
  }

  public startPortForward(): void {
    if (!this.config) {
      throw new Error('no config!');
    }
    this.log("Attempting to stop any existing port-forwards...");
    this.stopPortForward();
    this.log("Starting new port forwarding...");

    this.config.chains?.forEach(chain => {
      const chainPodPorts = this.podPorts.chains[chain.name] || this.podPorts.chains.defaultPorts;

      if (chain.cometmock?.enabled) {
        if (chain.ports?.rpc) this.forwardPortCometmock(chain, chain.ports.rpc, chainPodPorts.cometmock);
      } else {
        if (chain.ports?.rpc) this.forwardPort(chain, chain.ports.rpc, chainPodPorts.rpc);
      }
      if (chain.ports?.rest) this.forwardPort(chain, chain.ports.rest, chainPodPorts.rest);
      if (chain.ports?.grpc) this.forwardPort(chain, chain.ports.grpc, chainPodPorts.grpc);
      if (chain.ports?.exposer) this.forwardPort(chain, chain.ports.exposer, chainPodPorts.exposer);
      if (chain.ports?.faucet) this.forwardPort(chain, chain.ports.faucet, chainPodPorts.faucet);
    });

    this.config.relayers?.forEach(relayer => {
        const relayerPodPorts = this.podPorts.relayers[relayer.name] || this.podPorts.relayers.defaultPorts;
        if (relayer.ports?.rest) this.forwardPortRelayer(relayer, relayer.ports.rest, relayerPodPorts.rest);
        if (relayer.ports?.exposer) this.forwardPortRelayer(relayer, relayer.ports.exposer, relayerPodPorts.exposer);
    });

    if (this.config.registry?.enabled) {
      this.forwardPortService("registry", this.config.registry.ports.rest, this.podPorts.registry.rest);
      this.forwardPortService("registry", this.config.registry.ports.grpc, this.podPorts.registry.grpc);
    }

    if (this.config.explorer?.enabled) {
      this.forwardPortService("explorer", this.config.explorer.ports.rest, this.podPorts.explorer.rest);
    }
  }

  private getForwardPids(): string[] {
    const result = this.exec([
      "ps", "-ef",
      "|", "grep", "-i", "'kubectl port-forward'",
      "|", "grep", "-v", "'grep'",
      "|", "awk", "'{print $2}'"
    ]);
    const pids = (result || '').split('\n').map(pid => pid.trim()).filter(a => a !== '')
    return pids;
  }

  public stopPortForward(): void {
    this.log(chalk.green("Trying to stop all port-forward, if any...."));
    const pids = this.getForwardPids();
    pids.forEach(pid => {
      this.exec([
        "kill", "-15", pid
      ]);
    });
    this.exec(['sleep', '2']);
  }

  public printForwardPids(): void {
    const pids = this.getForwardPids();
    pids.forEach(pid => {
      console.log(pid);
    });
  }
}
