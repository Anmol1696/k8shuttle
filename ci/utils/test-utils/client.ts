import { relative } from 'path';
import strip from 'strip-ansi'

import { StarshipClient } from '../src'
import { config } from './config';

interface TestClientCtx {
  commands: string[];
  logs: string[];
  code: number;
}

export const createClient = () => {
  const ctx: TestClientCtx = {
    commands: [],
    logs: [],
    code: -1
  };

  const client = new StarshipClient({
    helmName: 'osmojs',
    helmFile: relative(process.cwd(), config.configPath),
    helmRepo: 'starship',
    helmRepoUrl: 'https://cosmology-tech.github.io/starship/',
    helmChart: 'devnet',
    helmVersion: 'v0.1.38'
  });

  const handler = {
    get(target: any, prop: string, _receiver: any) {
      const originalMethod = target[prop];
      if (typeof originalMethod === 'function') {
        return function(...args: any[]) {
          const argsString = args.map(arg => strip(JSON.stringify(arg))).join(', ');
          ctx.logs.push(`Call: ${prop}(${argsString})`);
          // @ts-ignore
          return originalMethod.apply(this, args);
        };
      }
      return originalMethod;
    }
  };

  const proxiedClient: StarshipClient = new Proxy(client, handler);

  // Overriding the exit method
  // @ts-ignore
  proxiedClient.exit = (code: number) => {
    ctx.code = code;
  };
  
  // Overriding the exec method
  // @ts-ignore
  proxiedClient.exec = (cmd: string[]) => {
    // @ts-ignore
    proxiedClient.checkDependencies();
    ctx.commands.push(cmd.join(' '));
    ctx.logs.push(cmd.join(' '));
  };

  // Overriding the log method
  // @ts-ignore
  proxiedClient.log = (cmd: string) => {
    ctx.logs.push(strip(cmd));
  };

  return {
    client: proxiedClient,
    ctx
  }
}


export const expectClient = (ctx: TestClientCtx, code: number) => {
  expect(ctx.logs.join('\n')).toMatchSnapshot();
  expect(ctx.commands.join('\n')).toMatchSnapshot();
  expect(ctx.code).toBe(code);
}
