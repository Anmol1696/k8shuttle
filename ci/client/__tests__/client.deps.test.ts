import { createClient, expectClient } from '../test-utils/client';

describe('StarshipClient', () => {
  it('missing deps', () => {
    const { client, ctx } = createClient();

    client.dependencies = client.dependencies.map(dep=>{
      if (['kubectl', 'docker'].includes(dep.name)) {
        return {
          ...dep,
          installed: false
        };
      }
      return dep;
    });

    // @ts-ignore
    client.exec(['something'])

    expectClient(ctx, 1);
  })
  it('has all deps', () => {
    const { client, ctx } = createClient();

    client.dependencies.forEach(dep => dep.installed = true);

    // @ts-ignore
    client.exec(['something'])

    expectClient(ctx, -1);
  })
});