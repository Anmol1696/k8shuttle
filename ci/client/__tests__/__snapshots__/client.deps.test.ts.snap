// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarshipClient has all deps 1`] = `
"Call: exec(["something"])
something"
`;

exports[`StarshipClient has all deps 2`] = `"something"`;

exports[`StarshipClient missing deps 1`] = `
"Call: exec(["something"])
Log⬇
✓kubectl
xdocker
✓helm
EndLog⬆
Log⬇

Please install the missing dependencies:
EndLog⬆
Log⬇


docker: https://docs.docker.com/get-docker/
For macOS, you may also consider Docker for Mac: https://docs.docker.com/desktop/install/mac-install/


EndLog⬆
something"
`;

exports[`StarshipClient missing deps 2`] = `"something"`;
